### ISSUE TITLE
make ExpectImpl subject type aware

### ISSUE THREAD

**Issue by robstoll** [robstoll/atrium-roadmap#33]
So far almost each method of ExpectImpl takes an `Expect<T>` and does something with it. For instance, if one wants to reuse `toBe` on a domain level, then one has to know that toBe works for `Any` type and thus is located under `ExpectImpl.any`. It would be better if one passes the `Expect<T>` upfront and we define extension methods on ExpectImpl as we do for Expect. Something along the line of:
```  
fun <T> Expect<T>.myToBe(t: T): Expect<T> = addAssertion(ExpectImpl(this).toBe(t))
```  

This has several advantages:
- simpler API as one only sees what is relevant, also the guidance is better
- Expect and ExpectImpl are very similar in its usage as well as in how they can be extended by library authors

disadvantage:
- ExpectImpl.builder does not make sense anymore. However, this is OK as it does not really belong here anyway as it is not an assertion function implementation but only a helper to define an assertion function.

**Comment by robstoll** [comment]
I would say we should even move to an extension based version for `ExpectImpl(this)`. Something like:
```
fun <T> Expect<T>.myToBe(t: T): Expect<T> = addAssertion(_domain.toBe(t))
```

**Comment by robstoll** [comment]
problem with an extension based approach is that we run quickly into ambiguity issues if we have subtypes which want to define an own _domain. For instance: Iterable <- Collection <- List

**Comment by robstoll** [comment]
Solution which should work, add an additional interface independent of Expect (e.g. AssertionContainer) add a method to Expect to turn the current into it. Define domain functions as extensions of it, retrieve impl from it and delegate to impl. Not much different than ServiceLoader but with scoping to Expect and woth default impl. One could still load an impl via ServiceLoader.

**Comment by robstoll** [comment]
I envision a syntax like the following:
```
fun <T> Expect<T>.myToBe(t: T): Expect<T> = _domain { addAssertion(myToBe(t)) }
```

**Comment by robstoll** [comment]
With this we will have just one domain module in the long run instead of api, robstoll, robstoll-lib and builders

**Comment by robstoll** [comment]
we have already started to transition away from ExpectImpl to atrium-logic with an extension methods/generative approach. 
It is also already included in v0.13.0 as experimental module and will probably stay experimental until 0.15.0
